### DISCUSSION TITLE
How get current page component from layout

### DISCUSSION THREAD

**Discussion by GioviQ** [dotnet/aspnetcore#56131] [Q&A]
Now in a Blazor layout to detect the current page I have to check the current uri with NavigationManager.
I would like to get access from the layout to the class name of the current Blazor component with page directive inside the body of the layout to see if it implements an interface. This is more general than checking an uri with string comparison.

**Comment by nithyav2017** [comment]
I work around the scenario. Here below the code.

PageTrackerServices Class:

```
 public class PageTrackerServices
 {
     public object CurrentPageInstance { get; set; }
     public event Action? OnPageChanged;

     public void SetCurrentPageInstance(object pageInstance)
     {
         CurrentPageInstance = pageInstance;
         OnPageChanged?.Invoke();
     }
 }
```

When call `PageTrackerServicesmethod.SetCurrentPageInstance(this) inside both  OnInitialized() and OnAfterRender(bool firstRender)  of the pages(Ex; `Home`, `Counter`).    it captures the current page instance and triggers the  `OnPageChanged` Event. This event notifies the `MainLayout` component. 

```
@code{
    protected override void OnInitialized()
    {
        PageTracker.SetCurrentPageInstance(this ); 
    }
    protected override void OnAfterRender(bool firstRender)
    {
        if (firstRender)
            PageTracker.SetCurrentPageInstance(this );
    }
}
```


MainLayout.Razer:

In the OnInitialized() event of MainLayout.razor,   OnPageChanged event is handled using InvokeAsync to switch to the UI thread.
```
@code
{
    private object currentPage;

    protected override void OnInitialized()
    {
         currentPage = PageTracker.CurrentPageInstance;
        
        PageTracker.onPageChanged += () =>
        {
            InvokeAsync(() =>
            {
                currentPage = PageTracker.CurrentPageInstance;
                StateHasChanged();  //re-render the UI
            });
        };
    }
}
```
Below is the code to display current page name and interfaces.
```

  <div class="sidebar" style="color:white">
      @if(currentPage != null)
      {
          <p>Current Page Type: @currentPage.GetType().Name</p>
          <p>Implemented Interfaces:</p>
          <ul>
              @foreach
              (var iface in currentPage.GetType().GetInterfaces())
              {
                  <li>@iface.Name</li>
              }
    
          </ul>
      }
      else
      {
          <p>No Current Page</p>
      }
  </div>

```

Hope this implementation align with your scenario.
